core/src/test/java/org/springframework/security/core/annotation/ExpressionTemplateSecurityAnnotationScannerTests.java

```diff
@@ -54,6 +54,43 @@ void parseMultipleMetaSourceAnnotationParameterWithAliasFor() throws Exception {
 		assertThat(preAuthorize.value()).isEqualTo("check(#name)");
 	}
 
+	@Test
+	void parseMetaSourceAnnotationWithEnumImplementingExpressionTemplateValueProvider() throws Exception {
+		Method method = MessageService.class.getDeclaredMethod("process");
+		PreAuthorize preAuthorize = this.scanner.scan(method, method.getDeclaringClass());
+		assertThat(preAuthorize.value()).isEqualTo("hasAnyAuthority(user.READ,user.WRITE)");
```

Comment:
The resulting expression may not be useful since SpEL may try and interpret `user` as a method on the root object. I think it should be:

```java
hasAnyAuthority('user.READ','user.WRITE')
```

which I believe you can achieve by placing single quotes in `Permission#getExpressionTemplateValue`

Reply:
Good catch. Even though there's no SpEL evaluation, always best to produce valid SpEL expressions. :)

Reply:
This is fixed now.